ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -349,6 +349,13 @@ let assert_host_has_iommu ~__context ~host =
   if List.assoc "iommu" chipset_info <> "true" then
     raise (Api_errors.Server_error (Api_errors.vm_requires_iommu, [Ref.string_of host]))
 
+(* Check if there are vgpus not allocated, since we may check the mem/vgpu after reservation *)
+let has_non_allocated_vgpus ~__context ~self =
```

Comment:
I think the solution is unreliable.
Consider the scenario if the VM has 3 VGPUs but only 1 of them is non-allocated, it will still get the calculation result we need double amount (6) VGPUs.
So I think we'd better fix this issue in the `assert_*` functions and only alloc PGPU/SR-IOV for that non-allocated.

Reply:
Thanks for comments. @krizex 
Consider the issue you mentioned before, but AFAIK that there is only one vgpu can be assigned to one VM, so the `list` at most get one element.
Yes, seems SR-IOV has the same issues, I'd like to have Rob's opinion on this, do we need to add checks on SR-IOV also to avoid double check? @robhoes

Reply:
Yes, the code is written so that in the future we could be able to support multiple VGPUs with not too much effort, but as of now a lot of places are already relying on the fact that there is a VGPU only. I think this is fine here as well

Reply:
Will add a comment above the function to remind that `assumption of one VGPU per VM`
Thanks